packages/trpc/server/routers/viewer/teams/inviteMember.handler.ts

```diff
@@ -22,106 +26,355 @@ type InviteMemberOptions = {
+    throw new TRPCError({
+      code: "BAD_REQUEST",
+      message: "You must provide at least one email address to invite.",
     });
+  }
 
-    if (input.isOrg && invitee) {
-      throw new TRPCError({
-        code: "NOT_FOUND",
-        message: `Email ${usernameOrEmail} already exists, you can't invite existing users.`,
-      });
-    }
+  return emailsToInvite;
+}
 
-    if (!invitee) {
-      // liberal email match
+async function getUserToInviteOrThrowIfExists({
+  usernameOrEmail,
+  orgId,
+  isOrg,
+}: {
+  usernameOrEmail: string;
+  orgId: number;
+  isOrg?: boolean;
+}) {
+  // Check if user exists in ORG or exists all together
+  const invitee = await prisma.user.findFirst({
+    where: {
+      OR: [{ username: usernameOrEmail, organizationId: orgId }, { email: usernameOrEmail }],
+    },
+  });
 
-      if (!isEmail(usernameOrEmail))
-        throw new TRPCError({
-          code: "NOT_FOUND",
-          message: `Invite failed because there is no corresponding user for ${usernameOrEmail}`,
-        });
+  // We throw on error cause we can't have two users in the same org with the same username
+  if (isOrg && invitee) {
+    throw new TRPCError({
+      code: "NOT_FOUND",
+      message: `Email ${usernameOrEmail} already exists, you can't invite existing users.`,
+    });
+  }
 
-      // valid email given, create User and add to team
-      await prisma.user.create({
-        data: {
-          email: usernameOrEmail,
-          invitedTo: input.teamId,
-          ...(input.isOrg && { organizationId: input.teamId }),
-          teams: {
-            create: {
-              teamId: input.teamId,
-              role: input.role as MembershipRole,
-            },
-          },
+  return invitee;
+}
+
+function checkInputEmailIsValid(email: string) {
+  if (!isEmail(email))
+    throw new TRPCError({
+      code: "BAD_REQUEST",
+      message: `Invite failed because ${email} is not a valid email address`,
+    });
+}
+
+async function createNewUserConnectToOrgIfExists({
+  usernameOrEmail,
+  input,
+  parentId,
+  connectionInfo,
+}: {
+  usernameOrEmail: string;
+  input: InviteMemberOptions["input"];
+  parentId?: number | null;
+  connectionInfo: ReturnType<typeof getOrgConnectionInfo>;
+}) {
+  const { orgId, autoAccept } = connectionInfo;
+
+  const createdUser = await prisma.user.create({
+    data: {
+      email: usernameOrEmail,
+      invitedTo: input.teamId,
+      organizationId: orgId && orgId, // If the user is invited to a child team, they are automatically added to the parent org
+      teams: {
+        create: {
+          teamId: input.teamId,
+          role: input.role as MembershipRole,
+          accepted: autoAccept, // If the user is invited to a child team, they are automatically accepted
         },
-      });
+      },
+    },
+  });
 
-      const token: string = randomBytes(32).toString("hex");
+  // We also need to create the membership in the parent org if it exists
+  if (parentId) {
```

Comment:
@roae @sean-brydon I'm not sure about double membership on child team members, do we have a doc spec for that?

Reply:
Im not entirely sure what you mean here - this is needed as ORGs are the "holder" for all members and thats where payments come from. If we create a ORG called 

Cal.com

And have a team called 

Marketing

Invite Designer to marketing -> if they are not in org we need to create that membership also due to it being the top level team